CorsixTH/Lua/entities/humanoids/staff/doctor.lua

```diff
@@ -182,14 +161,125 @@ function Doctor:updateSkill(consultant, trait, amount) -- luacheck: ignore 212
   end
 end
 
-function Doctor:trainSkill(consultant, trait, skill_thres, room_factor, staff_count)
-  -- TODO: tweak/rework this algorithm
-  -- TODO: possibly adjust based upon consultant's skill level?
-  --       possibly based on attention to detail?
-  local constant = 12.0
-  local staff_factor = constant + (staff_count-1)*(constant/6.0)
-  local delta = room_factor / (skill_thres * staff_factor)
-  self:updateSkill(consultant, trait, delta)
+--! Confirm doctor has specialism
+--!param trait (string) what to check (in is_xxx format)
+--!return true if specialised
+function Doctor:hasSpecialism(trait)
+  local specialisms = {
+    ["is_surgeon"] = self.profile.is_surgeon >= 1,
+    ["is_psychiatrist"] = self.profile.is_psychiatrist >= 1,
+    ["is_researcher"] = self.profile.is_researcher >= 1,
+  }
+  return specialisms[trait]
+end
+
+--! Generate the performance factor of the training room
+--!param doc_atd (number) Attention to detail of student
+--!param con_skill (number) Skill level of consultant
+--!param class_size (number) Total students in the training room
+--!return calculated performance
+function Doctor._calculateTrainingPerformance(doc_atd, con_skill, class_size)
+  -- Clamp a performance modifier's impact between 90% and 110%
+  local function clamp(value)
+    local base, extra = 0.9, 0.2 -- specify the +-10% variability
+    local factor = value * extra
+    return base + factor
+  end
+
+  local performers = {
+    doctor_focus = clamp(doc_atd),
+    teacher_ability = clamp(con_skill^2), -- ~1.0 modifier at 0.75 skill
+    random = math.t_random(0.9, 1, 1.1), -- Allow some randomness
+  }
+  -- Average the three performance modifiers to generate a single performance factor
+  local count, p_impact = 0, 0
+  for _, perf in pairs(performers) do
+    p_impact = p_impact + perf
+    count = count + 1
+  end
+  p_impact = p_impact / count
+
+  -- Class size impacts performance when there are more than 3 students at rate 2/n-1
+  local c_impact = class_size >= 3 and class_size - 1 or 2
+  c_impact = 2 / c_impact
+
+  return p_impact * c_impact
+end
+
+--! Calculates the amount of skill to learn towards becoming a doctor/consultant and
+--! any amount towards a specialist skill being taught
+--! There are detailed explanations on the logic to these calculations in the Wiki (see
+--! Training) - PR NOTE: If merge, make wiki page
+--!param consultant (entity) The person teaching the student
+--!param room_factor (number) What score training objects in the room give
+--!param student_count (number) Total number of students present
+function Doctor:trainSkills(consultant, room_factor, student_count)
+  --[[Information about these calculations based on TH:
```

Comment:
```suggestion
  --[[ Information about these calculations based on TH:
```

Reply:
Will be moved to wiki anyway so likely this bit may be deleted or condensed.